### ISSUE TITLE
Menu controls too fast (unusable) as of version 2.5.0

### ISSUE THREAD

**Issue by PronkMedia** [RetroPie/EmulationStation#200]
As of version 2.5.0 I cannot use emulationstation properly because scrolling through the menu's happens to fast. I was expecting that this was a known issue and would be fixed in version 2.5.1. but that is not the case.

It happens on a Pi3 when I control it via bluetooth (8bitdo) and USB (game controller and keyboard arrows). When I push up/down once the selected item will move to the top/bottom item in a gamelist almost instantly. I see it scrolling through the whole list but it's almost impossible to select the correct item. I can only navigate through the items one-by-one if I use a keyboard and push the arrow buttons very fast.

I have tried changing to different themes, different powersaving options and different transition types. This is not resolving the problem.

After downgrading to version 2.4.9 everything worked as expected again.

**Comment by hex007** [comment]
@PronkMedia Can you try the same with OtherSettings > Powersaver >> Disabled. 

Let me know if that solves your problem.

**Comment by PronkMedia** [comment]
@hex007 I did try that (just checked it again to be sure), it does not change anything. I also restarted after changing settings to be sure. I have tried to change all settings and different combination of settings, but it does not change this strange behaviour.

It seems like the whole interface is working at a very (too) high speed. Everything goes so fast that I can barely notice the carousel animation and transitions.

**Comment by hex007** [comment]
@PronkMedia  This might be unconventional but can you change the power supply to PC or Laptop 5V.

There are some cases where over-voltage of Pi causes erratic processing speed up. I have noticed it only once when i was testing a power supply.

**Comment by pjft** [comment]
I just tried it briefly and I can't really say I could replicate it using a
USB keyboard :S

Just to confirm, there wasn't a version 2.4.9, but rather 2.4.1 . Are you
downgrading to that one in particular?

Are you comfortable compiling code on your Pi?


- Paulo Tavares

**Comment by PronkMedia** [comment]
I indeed downgraded to 2.4.1. There is probably a better way to do it, but what I did was SSH into the Pi and entered the steps to do an install from source (checked the steps in the retropie-setup repo). The extra step I did was revert to the 2.4.1 commit instead of the latest commit of the master branch. Is there a better way to do this? If there is a quicker way, it makes it easier for me to test/switch versions.

I'm currently at work and do not have a lot of spare time. I'm going to test other power supplies asap but that may be this weekend earliest. I also know I use some overclock settings including over_voltage. I'm going to check what happens if I disable those overclock settings. But these settings were also present when using older Emulationstation versions.

I'm happy to do some extra (compile) steps/test if you can tell me what to do exactly.

**Comment by joolswills** [comment]
https://retropie.org.uk/forum/topic/12054/emulationstation-runs-too-fast-after-update

Can anyone reproduce this - those that have this problem - please can you give all details of set-up including OS, video output, video mode and all emulationstation configs from `/opt/retropie/configs/all/emulationstation`

We may need to temporarily revert the powersaving code if that is the cause of this (Or I can temporarily switch the ES build to an older version).

**Comment by pjft** [comment]
I'll try to test it out later today on my other build, but here on my dev build, with a keyboard, I can't seem to replicate it.

@PronkMedia I know this might be a bit of a pain, but if you would have the time and availability (very much appreciated, as I know you shouldn't need to spend time on this kind of endeavor), it would help if you would be available to help bisect it in order to narrow down the exact change that caused this.

If you - or anyone affected - would be willing to invest some time in this and help us out, it'd be very much appreciated.

Exit ES to the terminal, and then run the following (there, or via SSH):

Step 1:
```
git clone --recursive https://github.com/RetroPie/EmulationStation.git ES-bisect
cd ES-bisect
git bisect start
git bisect bad
git bisect good b28fa2ce9f6a4f9b0d8ad3af7bf09fc008c80d59
```

and then, the following:

Step 2:
```
cmake .
make
./emulationstation
```

and then test and exit ES.

Step 3:
After testing, if it works well, run

```
git bisect good
```

If it has the problem, run

```
git bisect bad
```

and then repeat steps 2 and 3, to create a new binary, and attempt to narrow down what exact change caused this.

It shouldn't take more than 3 repetitions of these, but it can still take a bit as compiling can take some time, so we definitely thank you in advance.

At the end it will tell you what commit caused it, so you can send it back to us.

Thank you.

**Comment by pjft** [comment]
@PronkMedia Actually, would you per chance have your ScreenSaver time set to zero? Can you check - either in ~/.emulationstation/es_settings.cfg, or in the menu?

Can you change it to another value - 60000 in the file / 1 min in the menu, or higher - and see if it fixes it?

**Comment by PronkMedia** [comment]
@pjft I checked the setting in ~/.emulationstation/es_settings.cfg (because the menu is unusable). It was 0. I have changed it to 60000 and after rebooting everything worked normally again. Than I changed it back to 0 and the problem was back after rebooting. So that is definitely causing the problem.

Disabling overclock settings, changing power source and different video output sources did not make any difference.

Do I still need to do the compiling steps? As this will take some time, I hopefully can find some time to do this in the coming weekend.

**Comment by pjft** [comment]
No need for compilation. This is the main issue. Thanks for confirming.

I do wonder: I don't think we changed the saving of that value, so the only
thing I can think of is that you had it set to zero all along?

Do you use a screensaver? Do you remember setting this value at any time?

Best.

**Comment by PronkMedia** [comment]
I'm 100% sure I have set it to 0 myself. I have disabled the screensaver by setting it to 0 minutes.

**Comment by pjft** [comment]
Correct. That is the expected behavior. Thanks.

**Comment by hex007** [comment]
@PronkMedia Can you set it back to something >0 and see if the problem persists?

**Comment by hex007** [comment]
@joolswills #199 has been updated to take care of this bug. 

@PronkMedia I would like to apologise for any inconvenience caused. Thanks for reporting this.

**Comment by joolswills** [comment]
I merges the commit. Thanks all. Please confirm a source build is ok, and I will bump version and rebuild bins later.

**Comment by hex007** [comment]
Pi 3 Builds Updated as of 3rd Aug 2017 `hex007:ps_updates`

https://www.dropbox.com/s/su7cpgvomy5edpx/emulationstation_3Aug_Pi3_ps_updates?dl=0

**Comment by hex007** [comment]
@joolswills Screensaver is not getting triggered now. This might take some time to fix.

**Comment by pjft** [comment]
Do you want me to test it out on my end, or are you looking into this?


- Paulo Tavares

**Comment by hex007** [comment]
@pjft Its late so I will mostly be looking at it tomorrow. I will certainly let you know when I need your help.

**Comment by pjft** [comment]
Sounds good - do have some rest!

I'm compiling it and will see if I can spot anything.

Thanks!


- Paulo Tavares

**Comment by joolswills** [comment]
I can revert it. I may revert all the power saving stuff if there is no quick fix. Please make it clear when opening a pr if not ready for merging

**Comment by joolswills** [comment]
Please let me know - if you have a fix today I can wait.

**Comment by joolswills** [comment]
Today or soon.

**Comment by pjft** [comment]
So, I spent a bit this morning compiling and tweaking the code, but this is indeed a bit more complex than ideal.

My initial recommendation is to revert the latest PR, and just push a change that addresses the "ScreensaverTime" != 0 bit, as well as keeping PowerSaver off for the video screensaver - which I have in my PR, so I can push that and add the screensaver time piece.

The reasons for that recommendation are that the current Power Savings handling effectively halts ES altogether, meaning that any logic within the different components (animations, changes, etc) that depends on "time" will be an edge case on its own. It doesn't seem to be simple to address the video screensaver behavior right now because of that.

Furthermore, this PR bundled a few refactorings together with a few fixes, and that makes it hard to narrow down what exactly causes the changes, and what works better now.

If people choose to have video screensaver On, they won't benefit from power savings for now, during its execution. That's fine. We can fix that later.

My two cents. 

--

Also, two things for future code tweaks:
* Part of the problem on particularly not launching the screensaver stems from the "-100" every now and then, as we're exiting the timeout before the screensaver detects that it's ready to launch (ssTime - 100). Unless there's a clear reason to do that "-100", let's not add those shortcuts. 
* Another slight change at the moment is that we ask it to timeout for longer than needed. We're currently asking it to timeout for (screensaver time), though it should be (screensaver time minus time elapsed). Just a note there.

Thanks!

**Comment by joolswills** [comment]
Happy if someone submits a PR with the revert + the suggested tweak only. I don't have much free time today.

**Comment by joolswills** [comment]
Or another PR that addresses the problems etc

**Comment by pjft** [comment]
I'll submit that then, and we can wait on @hex007 's feedback for what the next steps should be.

**Comment by joolswills** [comment]
@hex007 has a fix or open - probably should wait on that?

**Comment by pjft** [comment]
Correct. I'll just submit a PR for a revert and fix for 2.5.0, but we can wait to merge it depending on @hex007 's feedback on whether he has a fix in mind for the current changes. Our timezones must be different.

At least from my understanding of his feedback is that this may not be trivial to fix, and I certainly struggled as well. It's more of a "just in case it takes time, there's something in place for us to revert and keep going - otherwise we move forward with the new fixes".

I won't have time during the weekend, so just pre-empting that.

**Comment by pjft** [comment]
Sure, we can wait on @hex007 then.

**Comment by joolswills** [comment]
Re-opening this for now - I have forced ES source builds to v2.4.1 also until all related power saving issues are resolved. I think this is easier than reverting and having to re-do the code,

**Comment by pjft** [comment]
Thank you. Have a great weekend!

**Comment by hex007** [comment]
@pjft First things first Let me explain you the logic that is going on in my head so it is easier to understand the code. Comments added now are improperly indented to differentiate from code base
***
**PowerSaver.cpp**
```c++
int PowerSaver::getTimeout()
{
// If screensaver is running return PNtimeout else SStimeout
	return mRunningScreenSaver ? mPlayNextTimeout : mScreenSaverTimeout;
}

void PowerSaver::updateTimeouts()
{
	mScreenSaverTimeout = (unsigned int) Settings::getInstance()->getInt("ScreenSaverTime");
// If screensaver timeout is >0 wake up from sleep 100ms earlier so that screensaver is triggred normally. If SS is disabled using timeout==0 then wait till next event (-1)
	mScreenSaverTimeout = mScreenSaverTimeout > 0 ? mScreenSaverTimeout - 100 : -1;
// @pjft Change this line to get user set value from settings (currently 30secs)
	mPlayNextTimeout = 30000;
}
```

***
**main.cpp**
```c++
    // used to keep track of when the last event was received.
int ps_time = SDL_GetTicks();

bool running = true;
    // Used to keep track of when to shift to waiting
bool ps_standby = false;

while(running)
{
	SDL_Event event;
// go to standby if PS is enabled (Not user defined,ie. no backend logic needs frames to be updated. Safe to standby) and trailing time (user defined by PS_MODE) has elapsed since last event
	bool ps_standby = PowerSaver::getState() && SDL_GetTicks() - ps_time > PowerSaver::getMode();

	if(ps_standby ? SDL_WaitEventTimeout(&event, PowerSaver::getTimeout()) : SDL_PollEvent(&event))
	{
		do
		{
			switch(event.type)
			{
				//////////// And other Cases /////////////////
				case SDL_JOYDEVICEREMOVED:
					InputManager::getInstance()->parseEvent(event, &window);
					break;
				case SDL_QUIT:
					running = false;
					break;
			}
		} while(SDL_PollEvent(&event));

		// triggered if exiting from SDL_WaitEvent
// ps_standby is how we know inside this loop that we have reached here from wait and not poll
		if (ps_standby)
			// show as if continuing from last event
// This is used to smooth animations
			lastTime = SDL_GetTicks();

		// reset counter
// Update last event time
		ps_time = SDL_GetTicks();
	}

	if(window.isSleeping())
	{
		lastTime = SDL_GetTicks();
		SDL_Delay(1); // this doesn't need to be accurate, we're just giving up our CPU time until something wakes us up
		continue;
	}

	int curTime = SDL_GetTicks();
	int deltaTime = curTime - lastTime;
	lastTime = curTime;

	// cap deltaTime if it goes over non-zero Screensaver timeout or if negative
// Since we are taking out 100 in get timeout we add 100 here. 
// I think here we can just have deltaTime <0 condition
	if((deltaTime > PowerSaver::getTimeout() + 100 && PowerSaver::getTimeout() > 0) || deltaTime < 0)
		deltaTime = 1000;

	window.update(deltaTime);
	window.render();
	Renderer::swapBuffers();

	Log::flush();
}
```

**Comment by pjft** [comment]
Thanks.

What I experienced on the pi 3 was that a loop of the running cycle takes
17ms, so waking 100 ms before meant that they would check for start of
screensaver still before it's meant to be triggered, and then go to sleep
again for another timeout.

Waking up at any time after the screensaver timeout would trigger the
screensaver as expected, even if potentially a few milliseconds later.

That's why the -100 and +100 for me would do well in being removed.

**Comment by hex007** [comment]
I have found a sustainable solution and I am rewriting the main.cpp file's logic. I shall update the PR soon and then we can test this rigorously.

**Comment by hex007** [comment]
@joolswills @pjft Can we continue the discussion on #202

**Comment by hex007** [comment]
@joolswills  Considering this has been fixed, this issue can be closed now. New PR is about conflicts between PS and VideoPreview/SS which @pjft and I are working on getting perfected.
